### PR TITLE
Add canonical content export → source-controlled JSON + CI drift guard (#1418)

### LINKED ISSUE #1418: Content export → canonical source-controlled JSON (+ CI drift guard)
Part of the content-authoring-pipeline spike #1390 — decision 1, 2 & 4. See [`docs/spikes/1390-content-authoring-pipeline.md`](https://github.com/ginderjeremiah/GameServer/blob/main/docs/spikes/1390-content-authoring-pipeline.md).

Implements authoring model **(A) DB-primary + one-way export**: a deterministic exporter mirrors the static reference data to source-controlled JSON used only as a seed/snapshot. JSON is **generated, never hand-edited.**

## Scope
- A deterministic exporter that serializes the **9 static reference sets** — `Skill`, `Item`, `ItemMod`, `Enemy`, `Zone`, `Challenge`, `Class`, `Path`/`Proficiency`, `SkillRecipe` — to **one file per set** (`skills.json`, `items.json`, …). Source the existing **nested read contracts** (`Game.Abstractions.Contracts`), which already embed child collections (e.g. `Skill` embeds `DamageMultipliers`/`Effects`/`DamagePortions`).
- **Determinism rules** (the make-or-break for review-friendly diffs):
  - top-level array sorted by id (== index); **retired rows included** (they hold permanent slots).
  - child collections in a declared stable order (e.g. damage portions by `EDamageType`).
  - canonical number formatting (invariant culture; fixed `decimal`/`double` rendering so `1`/`1.0`/`1.00` don't churn).
  - `RetiredAt` normalized to stable UTC ISO-8601 (or null).
  - intrinsic/enum sets are **not** exported (they live in `HasData`/enums).
- **CI drift guard** (decision 4): a test that re-derives the export from a freshly-migrated-and-seeded DB and asserts byte-equality with the committed files, so a hand-edit or a stale snapshot fails the build.
- (Optional, nice-to-have) an admin "Export" affordance that writes the repo files.

## Notes / open
- Pick the export entry point: iterate each set's existing contract read, or add a thin aggregate exporter service. Keep it reusing the published read projection, not re-mapping entities.
- Decide the on-disk location (e.g. a `content/` folder) — this path is the contract the seed (#PLACEHOLDER_SEED) and lint (#PLACEHOLDER_LINT) consume.

## Tests
- Unit: ordering + canonical formatting are stable (re-export of the same data is byte-identical; field-order/number-format edge cases).
- The CI drift guard itself.

Blocks: the seed and lint sub-issues depend on the format this defines.

### PR DESCRIPTION
**Pull request by ginderjeremiah**
## Summary

Implements **decisions 2 & 4** of the content-authoring-pipeline spike (#1390) — authoring model **(A) DB-primary + one-way export**. A deterministic exporter mirrors the static reference-data graph to source-controlled JSON under `content/`, plus a CI drift guard that keeps the snapshot honest.

Closes #1418.

## What's included

- **`IContentExporter` / `ContentExporter`** (`Game.Application/Content`) — serializes the 10 static reference sets through the existing **nested read contracts** (`Game.Abstractions.Contracts`), i.e. the same published projection the client and Workbench receive, never the EF entities.
- **`ContentExportSerializer`** — the canonical, review-friendly, byte-stable format:
  - top-level arrays id-ordered (== cache index), retired rows included (permanent slots);
  - child collections in a **declared stable order** (e.g. damage portions by type, effects by id, attribute rows by attribute);
  - enums rendered as **names**, `RetiredAt` as a **UTC ISO-8601** instant (Unspecified treated as UTC for determinism);
  - fixed 2-space indent + **LF**, relaxed (verbatim) non-ASCII so prose reads naturally, trailing newline.
- **CI drift guard** (`ContentExportDriftGuardTests`) — re-derives the export from a freshly-migrated-and-seeded DB and asserts **byte-equality** with the committed files, so a hand-edit or a stale snapshot fails the build. `CONTENT_EXPORT_REGEN=1` rewrites the files (the documented "land a content change" workflow).
- **Committed `content/*.json` snapshot** — the export of the existing `e2e-seed.sql` content slice.
- **`ChallengeContractMapper`** — extracted the challenge domain→contract projection that was private to the `GetChallenges` socket command into a shared mapper, now reused by both the command and the exporter (DRY; behaviour-preserving).
- **`DatabaseCleaner` fix** — `Classes` wasn't truncated between integration tests (nothing cascades to it: `Players` references `Classes`, not the reverse), so class content leaked across tests. Added it (child tables cascade). This was surfaced by the drift guard failing only in the full suite.
- `.gitattributes` pins `content/*.json` to LF; `docs/backend.md` documents the export + drift-guard workflow.

## How it addresses the issue

The issue asks for a deterministic exporter over the 9 static sets sourced from the nested read contracts, with determinism rules and a CI drift guard. All delivered. Two interpretation notes worth a reviewer's eye:

1. **10 files, not 9.** `Path` and `Proficiency` are distinct reference sets with their own id-spaces (separate `AllPaths()` / `AllProficiencies()` reads), so each gets its own file (`paths.json` + `proficiencies.json`). The spike's literal file list omitted `proficiencies.json`; emitting both seemed clearly correct rather than cramming two id-spaces into one file. Easy to fold together if preferred.
2. **Seed source for the guard.** Since #1418 stands alone (the JSON-driven seeder is #1419), the drift guard seeds from the `e2e-seed.sql` content slice — the only source-controlled content today — so the committed snapshot is the export of that slice. This meaningfully catches a "stale snapshot" (edit `e2e-seed.sql` content, forget to re-export → build fails). #1419 reverses the direction (seed *from* `content/*.json`, retire `e2e-seed.sql`) and will repoint this guard's seed source.

## Test plan

- `dotnet build Game.sln` — clean (0 warnings/errors).
- `ContentExportSerializerTests` (6 unit tests) — ordering, idempotence, enum-name / UTC-ISO / LF / trailing-newline rendering, Unspecified-kind determinism.
- `ContentExportDriftGuardTests` — passes in assert mode against the committed files.
- Full `Game.Application.Tests` — **685 passed** (drift guard included, in the shared integration collection).
- Full `Game.Api.Tests` — **652 passed** (covers the `GetChallenges` refactor and shares `DatabaseCleaner`).

## Follow-ups

- The optional admin "Export" affordance (spike decision 4, nice-to-have) is **not** implemented; the `CONTENT_EXPORT_REGEN=1` test mode covers the regenerate workflow for now.
- Unblocks #1419 (seed a fresh DB from the export) and #1420 (progression-graph lint over the export), which consume the on-disk format defined here.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01CGknmb4DbRwJ5LmFfZp99b

---
_Generated by [Claude Code](https://claude.ai/code/session_01CGknmb4DbRwJ5LmFfZp99b)_

### CODE REVIEW

File: Game.Application/Content/ContentExportSerializer.cs [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+        }
+
+        public static IReadOnlyList<Contracts.Enemy> Canonicalize(IEnumerable<Contracts.Enemy> enemies)
+        {
+            var ordered = enemies.OrderBy(e => e.Id).ToList();
+            foreach (var enemy in ordered)
+            {
+                enemy.AttributeDistribution = enemy.AttributeDistribution.OrderBy(a => a.AttributeId).ToList();
+                enemy.SkillPool = enemy.SkillPool.OrderBy(s => s).ToList();
+                enemy.Spawns = enemy.Spawns.OrderBy(s => s.ZoneId).ToList();
```
Comment (ginderjeremiah) [inline Game.Application/Content/ContentExportSerializer.cs]: Nit (determinism robustness): `Spawns` is ordered by `ZoneId` alone, and `StarterEquipment` (line 116) by `EquipmentSlot` alone. These are total orders only as long as each parent has at most one row per key — true today, but if two spawns ever share a `ZoneId` (or two equipment rows a slot), the tie order falls back to cache materialization order and the bytes could churn. Every other collection here keys off a unique `Id`. Cheap insurance is a secondary `.ThenBy(...)` (e.g. `.ThenBy(s => s.EnemyId)` / by weight), so the ordering is provably total. Not blocking — just flagging while the format is still cheap to harden.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: Game.Application/Content/ContentExportSerializer.cs [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+        /// <summary>
+        /// Writes a <see cref="DateTime"/> as a stable UTC ISO-8601 instant so a retire timestamp's serialized
+        /// form never depends on the reader's locale or the value's <see cref="DateTimeKind"/>. An Unspecified
+        /// kind (the shape a bare <c>timestamp</c> column reads back as) is taken to already be UTC.
+        /// </summary>
+        private sealed class UtcIsoDateTimeConverter : JsonConverter<DateTime>
+        {
+            private const string Format = "yyyy-MM-dd'T'HH:mm:ss.fffffff'Z'";
+
+            public override DateTime Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
```
Comment (ginderjeremiah) [inline Game.Application/Content/ContentExportSerializer.cs:165]: Nit: `Read` is currently unused — the exporter only writes, and nothing deserializes this format yet. It's reasonable to keep a converter symmetric, and #1419's seeder will consume it shortly, but per the no-dead-code guideline it's worth a one-line note that `Read` exists for the upcoming seed path (or land it with #1419) so a future reader doesn't flag it as orphaned. Take it or leave it.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by ginderjeremiah — COMMENTED** [review]
## Review: Canonical content export → source-controlled JSON + CI drift guard (#1418)

**Verdict:** Approve _(submitted as a comment since I'm the PR author; this is a genuine approve, not a deferral)_ — clean, well-layered, and the determinism story holds up under scrutiny. Only minor nits, none blocking.

I focused the review on the two things most likely to bite this feature: the in-place mutation in `Canonicalize`, and cross-platform byte-stability of the export (the whole point of the drift guard).

### Correctness — verified, no issues
- **In-place reordering is safe.** `Canonicalize` reassigns child collections (`skill.DamageMultipliers = …OrderBy().ToList()`) on the contract objects. I traced every set the exporter reads (`AllSkills`/`All`/`AllPaths`/`AllProficiencies`/`AllSkillRecipes`): each maps fresh contracts per call via `.Select(Mapper.ToContract)`, and the mappers build fresh child lists with `.Select(…).ToList()`. So nothing here mutates a cached snapshot graph — the "freshly mapped per read" comment is accurate. `Challenges.All()` does return the shared cached list, but the exporter maps it through `ChallengeContractMapper.ToContract` first and the challenge `Canonicalize` doesn't touch child collections, so that path is clean too. Good catch keeping this off the `backend.md` cached-`Include`-graph footgun.
- **Determinism is solid.** Every ordering key is an `int` id or an enum (`OrderBy` on enums compares the underlying value) — no culture-sensitive string comparisons anywhere (`Tags` are `int` ids, not names), so the export is machine-independent. The explicit invariant-culture/UTC handling on the one `DateTime` field is the right call.

### Layering / DDD
Correct: the exporter lives in `Game.Application/Content`, reads the published read contracts (not entities), is registered scoped, and the `GetChallenges` projection was extracted to a shared `ChallengeContractMapper` (DRY, behaviour-preserving) rather than duplicated. No domain logic leaked anywhere.

### Nits (non-blocking)
- Single-key child orderings (`Enemy.Spawns` by `ZoneId`, `Class.StarterEquipment` by `EquipmentSlot`) are total orders only while each parent has ≤1 row per key. A secondary `.ThenBy(...)` would make them provably total and tie-proof. (inline)
- `UtcIsoDateTimeConverter.Read` is currently unused (export only writes); fine to keep for symmetry / #1419, worth a one-line note. (inline)
- Per-set `Canonicalize` ordering is unit-tested directly only for `Skill`; the other sets are covered solely by the drift guard, and most are empty in today's seed slice. The logic is uniform so this is low-risk, but a couple of representative cases (e.g. an `Item` with `Tags`/`ModSlots`, a `Proficiency` with `LevelModifiers`) would lock the ordering down independent of the seed.

### Test coverage
Adequate per the project's rules. Serializer determinism rules have focused unit tests (ordering, idempotence, enum-name/UTC-ISO/LF/trailing-newline, Unspecified-kind), and the integration-level drift guard re-derives from a migrated+seeded DB and asserts byte-equality. No frontend/battle-logic surface, so no mirrored-test obligation. The `DatabaseCleaner` `Classes` fix is a legitimate correctness catch surfaced by the guard.

### What's good
The two interpretation calls in the PR description (10 files not 9 — `paths`/`proficiencies` are distinct id-spaces; seeding the guard from the `e2e-seed.sql` slice until #1419 reverses direction) are the right judgment calls and clearly documented. The `backend.md` write-up and the inline rationale on the determinism/Unspecified-kind decisions are exactly the right level of detail. Solid, self-contained foundation for #1419/#1420.

I'll hold off merging: `test-backend` is still running, and there are open nits — once it's green you can address or wave off the nits and merge.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by ginderjeremiah** [comment]
Pushed `fe53279` addressing the review and the CI failure:

- **Total ordering nit** — `Enemy.Spawns` now `ThenBy(Weight)` and `Class.StarterEquipment` `ThenBy(ItemId)`, so a future duplicate key can't fall back to cache order. The committed `content/*.json` are unchanged (the seed slice has ≤1 spawn per zone and no starter equipment, so the tie-breaks don't reorder anything).
- **`Read` nit** — added a one-line note that it exists for the #1419 seed path, and a round-trip unit test now exercises it.
- **Per-set ordering coverage** — added direct `Canonicalize` unit tests for `Item`/`ItemMod`/`Enemy`/`Class`/`Proficiency`/`SkillRecipe` (the sets that are empty in today's seed slice), plus `Local`-kind `DateTime` handling.

These also fix the **coverage gate**: the new tests bring `Game.Application` back to 96.8% line / 91.7% branch locally (floors 95/85).

On the other red check: `Game.Infrastructure.Tests` reported 1 failure in the first run, but it passes **60/60** locally and again under the full merged-coverage run, and nothing in this PR touches that project — it looks like an unrelated flake. This push re-runs CI; I'll keep watching and dig in if it recurs.

---
_Generated by [Claude Code](https://claude.ai/code)_
